[Doc] Refactor simt (#6151)
taichi-dev/taichi: docs/lang/articles/reference/simt.md

---
sidebar_position: 5
---

# SIMT intrinsics

> For CUDA backend, Taichi now supports warp-level and block-level intrinsics that

Reviewer: ```suggestion
For the CUDA backend, Taichi supports warp-level and block-level intrinsics, which
```

are needed for writing high-performance SIMT kernels. You can use them in Taichi
similar to the [usage in CUDA kernels](https://developer.nvidia.com/blog/using-cuda-warp-level-primitives/). Currently, the following functions are supported:

| Operation                  | Mapped CUDA intrinsic         |
| -------------------------- | ----------------------------- |
|`ti.simt.warp.all_nonzero`  | `__all_sync`      |
|`ti.simt.warp.any_nonzero`  | `__any_sync`      |
|`ti.simt.warp.unique`       | `__uni_sync`      |
|`ti.simt.warp.ballot`       | `__ballot_sync`   |
|`ti.simt.warp.shfl_sync_i32`| `__shfl_sync`     |
|`ti.simt.warp.shfl_sync_f32`| `__shfl_sync`     |
|`ti.simt.warp.shfl_up_i32`  | `__shfl_up_sync`  |
|`ti.simt.warp.shfl_up_f32`  | `__shfl_up_sync`  |
|`ti.simt.warp.shfl_down_i32`| `__shfl_down_sync`|
|`ti.simt.warp.shfl_down_f32`| `__shfl_down_sync`|
|`ti.simt.warp.shfl_xor_i32` | `__shfl_xor_sync` |
|`ti.simt.warp.match_any`    | `__match_any_sync`|
|`ti.simt.warp.match_all`    | `__match_all_sync`|
|`ti.simt.warp.active_mask`  | `__activemask`    |
|`ti.simt.warp.sync`         | `__syncwarp`      |

> Please refer to our [API docs](https://docs.taichi.graphics/api/taichi/lang/simt/warp/#module-taichi.lang.simt.warp)

Reviewer: ```suggestion
See [Taichi's API reference](https://docs.taichi-lang.org/api/taichi/lang/simt/warp/#module-taichi.lang.simt.warp)
```

for more information on each function.

> Here is an example to perform data exchange within a warp in Taichi:

Reviewer: ```suggestion
Here is an example of performing data exchange within a warp in Taichi:
```
